lib/internal/repl/import.js

```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+const acorn = require('internal/deps/acorn/dist/acorn');
+const ESLoader = require('internal/loader/singleton');
+
+const kSymbolKey = 'internalReplLoaderPromise';
+const kInternalLoaderPromise = Symbol.for(kSymbolKey);
+
+/*
+Performs importing from static import statments
+
+"import { x, y } from 'some_modules'; x(); y"
+
+would become:
+
+(async () => {
+  const s = Symbol.for('internalReplLoaderPromise');
+  await this[s];
+  delete this[s];
+})().then(() => { x(); return (y); });
+
+The symbol is a promise from Promise.all on an array of promises
+from Loader#import. When a promise finishes it attaches to
+`repl.context` with a getter, so that the imported names behave
+like variables not values (live binding).
+*/
+
+
+function processTopLevelImport(src, repl) {
+  let root;
+  try {
+    root = acorn.parse(src, { ecmaVersion: 8, sourceType: 'module' });
+  } catch (err) {
+    return null;
+  }
+
+  const importLength =
+    root.body.filter((n) => n.type === 'ImportDeclaration').length;
+
+  if (importLength === 0)
+    return null;
+
+  const importPromises = [];
+  let newBody = '';
+
+  for (const index in root.body) {
+    const node = root.body[index];
+    if (node.type === 'ImportDeclaration') {
+      const lPromise = ESLoader.loader.import(node.source.value)
+        .then((exports) => {
+          const { specifiers } = node;
+          if (specifiers[0].type === 'ImportNamespaceSpecifier') {
+            Object.defineProperty(repl.context, specifiers[0].local.name, {
+              get() { return exports; }
+            });
+          } else {
+            const properties = {};
+            for (const { imported, local } of specifiers) {
+              const imp = imported ? imported.name : 'default';
+              properties[local.name] = {
+                get() { return exports[imp]; }
+              };
+            }
+            Object.defineProperties(repl.context, properties);
+          }
+        });
+
+      importPromises.push(lPromise);
+    } else if (node.expression !== undefined) {
+      const { start, end } = node.expression;
+      if (+index === root.body.length - importLength)
+        newBody += `return (${src.substring(start, end)});`;
+      else
+        newBody += src.substring(start, end) + ';';
+    } else {
+      newBody += src.substring(node.start, node.end);
+    }
+  }
+
+  repl.context[kInternalLoaderPromise] = Promise.all(importPromises);
+
+  return `(async () => {
+  const s = Symbol.for('${kSymbolKey}');
+  await this[s];
+  delete this[s];
+})().then(() => { ${newBody} });`;
```

Comment:
What if you just returned
```js
{
  importPromises,
  codeWithoutImports: newBody,
}
```
here, and then did something like this in `defaultEval`:
```js
if (!wrappedCmd && experimentalModules && /\bimport\b/.test(code)) {
  const processed = processTopLevelImport(code, self);
  if (processed) {
    Promise.all(processed.importPromises).then(() => {
      // Call defaultEval again once the import promises have resolved:
      defaultEval(processed.codeWithoutImports, context, file, cb);
    });
    return;
  }
}
```
Then you wouldn't have to generate any special code to `await` the promises, and you wouldn't have to set `repl.context[kInternalLoaderPromise]`.

Reply:
thats an excellent idea, thank you 😄